src/main/java/pl/javastart/task/PriceController.java

```diff
@@ -0,0 +1,133 @@
+import java.util.List;
+
+public class PriceController {
+    private static final int ROUNDING_SCALE = 10;
+    private static final int ROUNDING_MODE = 0;
+
+    public static List<Product> readProducts(String fileName) {
+        List<Product> products = new ArrayList<>();
+        try (var br = new BufferedReader(new FileReader(fileName))) {
+            String nextLine;
+            while ((nextLine = br.readLine()) != null) {
+                Product product = parseProduct(nextLine);
+                products.add(product);
+            }
+            return products;
+        } catch (IOException ex) {
+            System.err.println("Wystąpił problem podczas czytania pliku");
+            return List.of();
+        }
+    }
+
+    private static Product parseProduct(String line) {
+        String[] lineAsArray = line.split(";");
+        String name = lineAsArray[0];
+        BigDecimal price = new BigDecimal(lineAsArray[1]);
+        String currency = lineAsArray[2];
+        return new Product(name, price, currency);
+    }
+
+    public static List<Currency> readCurrencies(String fileName) {
+        List<Currency> currencies = new ArrayList<>();
+        try (var br = new BufferedReader(new FileReader(fileName))) {
+            String nextLine;
+            while ((nextLine = br.readLine()) != null) {
+                Currency currency = parseCurrency(nextLine);
+                currencies.add(currency);
+            }
+            return currencies;
+        } catch (IOException ex) {
+            System.err.println("Wystąpił problem podczas czytania pliku");
+            return List.of();
+        }
+    }
+
+    private static Currency parseCurrency(String line) {
+        String[] lineAsArray = line.split(";");
+        String name = lineAsArray[0];
+        BigDecimal rate = new BigDecimal(lineAsArray[1]);
+        return new Currency(name, rate);
+    }
+
+    public static BigDecimal getSumInEuro(List<Product> products, List<Currency> currencies) {
+        BigDecimal sum = BigDecimal.ZERO;
+        for (Product product : products) {
+            BigDecimal rate = findRate(product.getCurrency(), currencies);
+            BigDecimal euroPrice = product.getPrice()
+                    .divide(rate, ROUNDING_SCALE, ROUNDING_MODE);
+            sum = sum.add(euroPrice);
+        }
+        return sum;
+    }
+
+    public static BigDecimal getAveragePriceInEuro(List<Product> products, List<Currency> currencies) {
+        BigDecimal sum = getSumInEuro(products, currencies);
+        return sum.divide(new BigDecimal(products.size()), ROUNDING_SCALE, ROUNDING_MODE);
+    }
+
+    public static Product getMostValuableProduct(List<Product> products) {
+        if (products.size() == 0) {
+            throw new EmptyFileException("Brak produktów");
+        }
+        Product mostValuable = products.get(0);
+        for (Product product : products) {
+            if (product.getPrice().compareTo(mostValuable.getPrice()) > 0) {
```

Comment:
To porównanie jest z założenia niepoprawne, bo produkty są w różnych walutach. Tutaj powinnaś przeliczać produkt z waluty oryginalnej na EUR, żeby porównywać jabłka do jabłek